compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -4832,15 +4843,107 @@ private STNode parseFuncCall(STNode identifier) {
      * Parse error constructor expression.
      * </p>
      * <code>
-     * error-constructor-expr := error ( arg-list )
+     * error-constructor-expr := error [error-type-reference] ( error-arg-list )
+     * error-arg-list := positional-arg [, positional-arg] (, named-arg)*
      * </code>
      *
      * @return Error constructor expression
      */
     private STNode parseErrorConstructorExpr() {
+        startContext(ParserRuleContext.ERROR_CONSTRUCTOR);
         STNode errorKeyword = parseErrorKeyword();
-        errorKeyword = createBuiltinSimpleNameReference(errorKeyword);
-        return parseFuncCall(errorKeyword);
+        STNode typeReference = parseErrorTypeReference(errorKeyword, ParserRuleContext.ERROR_CONSTRUCTOR_RHS);
+        STNode openParen = parseOpenParenthesis(ParserRuleContext.ARG_LIST_OPEN_PAREN);
+        return parseErrorConstructorExpr(errorKeyword, typeReference, openParen);
+    }
+
+    private STNode parseErrorTypeReference(STNode errorKeyword, ParserRuleContext context) {
+        STToken nextToken = peek();
+        switch (nextToken.kind) {
+            case IDENTIFIER_TOKEN:
+                return parseTypeReference();
+            case OPEN_PAREN_TOKEN:
+                return STNodeFactory.createEmptyNode();
+            default:
+                recover(nextToken, context, errorKeyword);
+                return parseErrorTypeReference(errorKeyword, context);
+        }
+    }
+
+    private STNode parseErrorConstructorExpr(STNode errorKeyword, STNode typeReference, STNode openParen) {
+        STNode functionArgs = parseArgsList();
+        STNode errorArgs = getErrorArgList(functionArgs);
+        STNode closeParen = parseCloseParenthesis(ParserRuleContext.ARG_LIST_CLOSE_PAREN);
+        endContext();
+
+        openParen = cloneWithDiagnosticIfListEmpty(errorArgs, openParen,
+                DiagnosticErrorCode.ERROR_MISSING_POSITIONAL_ARG_IN_ERROR_CONSTRUCTOR);
+        return STNodeFactory.createErrorConstructorExpressionNode(errorKeyword, typeReference, openParen, errorArgs,
+                closeParen);
+    }
+
+    /**
+     * Validate function args and return error arg node list.
+     *
+     * @param functionArgs function args to be validated
+     * @return Error arg list
+     */
+    private STNode getErrorArgList(STNode functionArgs) {
+        STNodeList argList = (STNodeList) functionArgs;
+        if (argList.isEmpty()) {
+            return argList;
+        }
+
+        List<STNode> errorArgList = new ArrayList<>();
+        // Validate first arg
+        STNode arg = argList.get(0);
+        switch (arg.kind) {
+            case POSITIONAL_ARG:
+                errorArgList.add(arg);
+                break;
+            case NAMED_ARG:
+                arg = SyntaxErrors.addDiagnostic(arg,
+                        DiagnosticErrorCode.ERROR_MISSING_POSITIONAL_ARG_IN_ERROR_CONSTRUCTOR);
+                errorArgList.add(arg);
+                break;
```

Comment:
We need to restrict positional argument after named arg right? ie what if `(namedArg, possitionalArg)` ? I think we can use `hasPositionalArg` boolean for this with a better name like `allowPositionalArg`

Reply:
What we do here is validating function args. That order validation is already done when parsing function args.

Reply:
ah cool.